LSPS4/README.md

```diff
@@ -0,0 +1,1923 @@
+> single batched transaction).
+> This is done by `error`ing the channel before broadcasting the
+> funding transaction, then re-starting the channel open flow
+> with the failing peer removed from the batched channel funding
+> transaction.
+> Thus, other clients will see their channel opens `error`ed
+> if one of the clients in the same batch aborts the flow, but as
+> long as the LSP attempts a re-open before continuing with
+> payment, this is not a funds loss event, and neither the LSP nor
+> the client is obligated to pay onchain fees in this case, only
+> spend CPU processing time and network bandwidth on the additional
+> negotiations.
+
+Depending on the `client_trusts_lsp` from the `lsps4.get_opening_fee`
+result, after the client has sent `funding_signed` and the LSP is
+willing to broadcast the funding transaction:
+
+* If `client_trusts_lsp` is specified and `true`:
+  * The LSP MAY wait for the client to send the preimages to the
+    incoming HTLCs (or equivalent for other PPP types) via an
+    `update_fulfill_htlc` before broadcasting the funding
+    transaction.
+  * The client MUST NOT wait for the LSP to broadcast the funding
+    transaction before sending the preimage via a
+    `update_fulfill_htlc`.
+  * The client and LSP MUST immediately send `channel_ready`.
+* If `client_trusts_lsp` is unspecified or `false`:
+  * The client MAY wait for the funding transaction to appear in
+    its mempool or the mempool of a trusted node, or confirmed in a
+    block with any depth, AND confirm that the funding transaction
+    output has the correct amount and `scriptPubKey`, before sending
+   `channel_ready`.
+  * The LSP MUST immediately broadcast the channel funding
+    transaction and send `channel_ready`.
+  * The client MAY immediately send `channel_ready` (i.e. the
+    client trusts the LSP, even though the LSP does not require
+    that trust).
+
+> **Rationale** As mentioned in the section "Trust Models" above,
+> in LSP trusts client model, the client is allowed to choose its
+> risk appetite; it may accept 0 confirmations, 1, 3, 6, 100, or
+> however many it believes would be safe with the particular LSP.
+
+As soon as both client and LSP have sent `channel_ready` for the
+new channel, the LSP then forwards PPPs in the PPP Opening Set to
+the client.
+
+The LSP MUST NOT forward other HTLCs to the new channel until
+the PPP Opening Set is empty.
+
+The LSP generates non-standard forwards, where the amount received
+by the client is smaller than specified in the onion;
+the client MUST accept the non-standard forward(s), provided they
+sum to at least `payment_size_msat - opening_fee`.
+
+The LSP MUST ensure that each forwarded part forwards at least
+`client_htlc_minimum_msat` millisatoshis.
+The LSP MUST ensure that all parts forwarded sum up to at least
+`payment_size_msat - opening_fee`.
+
+The LSP MUST include an `extra_fee` (type 65536) TLV to the
```

Comment:
Next to the `extra_fee` TLV, every lsps4 `update_add_htlc` probably needs a `ppp_id` TLV, to correlate the actual HTLC with the PPP.

Reply:
Why though?  The PPP exists only as a way to implement onion oracles (where the oracle is the client). In the end an HTLC is an HTLC.